src/uct/cuda/cuda_copy/cuda_copy_md.c

```
@@ -221,10 +221,11 @@ uct_cuda_copy_mem_alloc_fabric(uct_cuda_copy_md_t *md,
         return status;
     }
 
-    prop.type                 = CU_MEM_ALLOCATION_TYPE_PINNED;
-    prop.requestedHandleTypes = CU_MEM_HANDLE_TYPE_FABRIC;
-    prop.location.type        = CU_MEM_LOCATION_TYPE_DEVICE;
-    prop.location.id          = cu_device;
+    prop.type                            = CU_MEM_ALLOCATION_TYPE_PINNED;
+    prop.requestedHandleTypes            = CU_MEM_HANDLE_TYPE_FABRIC;
+    prop.location.type                   = CU_MEM_LOCATION_TYPE_DEVICE;
+    prop.location.id                     = cu_device;
+    prop.allocFlags.gpuDirectRDMACapable = 1;
```

Will the allocation fail on a system with no IB?

---

@SeyedMir 
No. `gpuDirectRDMACapable` is a request. CUDA isn't obliged to honor it and report errors if rdma enablement fails. Instead user is expected to check attribute `CU_POINTER_ATTRIBUTE_IS_GPU_DIRECT_RDMA_CAPABLE` once allocation succeeds to see if RDMA capability is enabled. 

But thanks for raising this as we do need to check that allocation has handle fabric allowed with it and we should return a failure if that isn't possible. I've made relevant changes now.

---

What about systems without RDMA (i.e TCP+shared mem only), maybe passing such flag could lead to failure?

---

Maybe we should not fail if the allocation is not RDMA capable, and let it fail later if actually used with RDMA

---

1. does it mean users have to pass this flag if they allocate fabric memory and pass it to UCX/MPI?
2. why it doesn't fail on new systems?

---

1. Indeed they are expected to pass a flag during allocation call
2. Under what scenarios do the tests currently fail? In general, passing a flag is a request to the allocator. The allocation doesn't fail if the request was made and the allocator wasn't able to honor the request. It is only at the point of trying to register that we are able to detect if the buffer is RDMA-capable.

---

> What about systems without RDMA (i.e TCP+shared mem only), maybe passing such flag could lead to failure?

It should not cause a failure.

> Maybe we should not fail if the allocation is not RDMA capable, and let it fail later if actually used with RDMA

We wouldn't fail the allocation and I agree that we should just fail at the point of registration.

---

> 2\. Under what scenarios do the tests currently fail? In general, passing a flag is a request to the allocator. The allocation doesn't fail if the request was made and the allocator wasn't able to honor the request. It is only at the point of trying to register that we are able to detect if the buffer is RDMA-capable.

Well, today we don't pass RDMA flag to the allocation, and CI passes, does it mean we never tried to register VMM rdnv fragments with IB? Or the issue would be exposed on specific type of system/GPU/Cuda.. ?

---

This should show up with a VMM allocation on any platform. If we try to register device VMM memory and the flag is not set by caller, then IB registration call should return a failure.